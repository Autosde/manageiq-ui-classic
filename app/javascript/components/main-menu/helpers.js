export const getHrefByType = (type, href, id) => {
  switch (type) {
    case 'big_iframe':
      return `/dashboard/iframe?id=${id}`;
    case 'modal':
      return undefined;
    default:
      return href;
  }
};

export const getTargetByType = type => (type === 'new_window' ? '_blank' : '_self');

export const handleUnsavedChanges = (e, type) => {
  if (type === 'modal') {
    return sendDataWithRx({ type: 'showAboutModal' });
  }
  return window.miqCheckForChanges();
};
