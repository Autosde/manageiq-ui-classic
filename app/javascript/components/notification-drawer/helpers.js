export const newCountText = function(count) {
  return sprintf(n__('%d New', '%d New'), count);
};

export const getNotficationStatusIconName = notification => (
  {
    info: 'info',
    error: 'error-circle-o',
    danger: 'error-circle-o',
    warning: 'warning-triangle-o',
    success: 'ok',
    ok: 'ok',
  }[notification.type] || '');

export const viewDetails = function(notification) {
  window.location.href = `/restful_redirect?${$.param(notification.data.link)}`;
};
