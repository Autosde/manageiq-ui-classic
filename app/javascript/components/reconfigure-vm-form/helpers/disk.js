import { TYPES, sizeInMB } from './general';

/** Function to execute when the delete backing toggle changed */
export const backupDelete = (id, data, setData) => {
  let backingIds = data.submitParams.diskIds.backup || [];
  const index = id.substring(4);
  if (backingIds.includes(id)) {
    // code to remove from delete backlog
    data.dataTable.disks[index].delete_backing = false;
    backingIds = backingIds.filter((diskId) => diskId !== id);
  } else {
    // code to add the disk to delete backlog
    data.dataTable.disks[index].delete_backing = true;
    backingIds.push(id);
  }
  setData({
    ...data,
    dataTable: {
      ...data.dataTable,
      disks: data.dataTable.disks,
    },
    submitParams: {
      ...data.submitParams,
      diskIds: {
        ...data.submitParams.diskIds,
        backup: backingIds,
      },
    },
  });
};

/** function to delete disk executed when the delete button gets clicked for the disk datatable */
export const deleteDisk = (row, data, setData) => {
  let dataTableData = data.dataTable.disks;
  let addDisks = data.submitParams.disks.add || [];
  const deleteDisks = data.submitParams.disks.delete || [];
  let addDiskIds = data.submitParams.diskIds.add || [];
  const deleteIds = data.submitParams.diskIds.delete || [];

  if (addDiskIds.includes(row.id)) {
    // delete from the newly added disks
    // eslint-disable-next-line no-alert
    if (window.confirm(__('Are you sure you want to delete this entry?'))) {
      dataTableData = data.dataTable.disks.filter((disk) => disk.id !== row.id);
      addDisks = data.submitParams.disks.add.filter((disk) => disk.id !== row.id);
      addDiskIds = data.submitParams.diskIds.add.filter((diskId) => diskId !== row.id);
      data.changeCounter -= 1;
    }
  } else {
    // remove from the existing disk datatable
    const index = row.id.substring(4);
    const currentRow = data.dataTable.disks[index];
    currentRow.add_remove = 'remove';
    const deleteDiskData = {
      id: row.id,
      disk_name: currentRow.hdFilename,
      delete_backing: deleteIds.includes(row.id),
    };
    deleteDisks.push(deleteDiskData);
    deleteIds.push(row.id);
    data.changeCounter += 1;
  }
  setData({
    ...data,
    dataTable: {
      ...data.dataTable,
      disks: dataTableData,
    },
    submitParams: {
      ...data.submitParams,
      disks: {
        ...data.submitParams.disks,
        add: addDisks,
        delete: deleteDisks,
      },
      diskIds: {
        ...data.submitParams.diskIds,
        add: addDiskIds,
        delete: deleteIds,
      },
    },
  });
};

/** Cancel delete disk function
 *  button click from datatable
 */
export const cancelDeleteDisk = (row, data, setData) => {
  const submitDisks = data.submitParams.disks.delete.filter((disk) => disk.id !== row.id);
  const deletedDiskIds = data.submitParams.diskIds.delete.filter((id) => id !== row.id);
  data.changeCounter -= 1;
  setData({
    ...data,
    submitParams: {
      ...data.submitParams,
      disks: {
        ...data.submitParams.disks,
        delete: submitDisks,
      },
      diskIds: {
        ...data.submitParams.diskIds,
        delete: deletedDiskIds,
      },
    },
  });
};

/** Cancel resize disk function - restore the previous data size
 *  button click from datatable
 */
export const cancelResizeDisk = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(4);
  data.dataTable.disks[index].hdSize = data.dataTable.disks[index].old_size;
  data.dataTable.disks[index].hdUnit = data.dataTable.disks[index].old_unit;
  const submitDisks = data.submitParams.disks.resize.filter((disk) => disk.id !== row.id);
  const resizeDiskIds = data.submitParams.diskIds.resize.filter((id) => id !== row.id);
  data.changeCounter -= 1;
  setData({
    ...data,
    submitParams: {
      ...data.submitParams,
      disks: {
        ...data.submitParams.disks,
        resize: submitDisks,
      },
      diskIds: {
        ...data.submitParams.diskIds,
        resize: resizeDiskIds,
      },
    },
  });
};

/** Function to check the button need to be disabled or not */
const diskDisableButton = (id, data) => {
  if (data.submitParams.diskIds.delete.includes(id) || data.submitParams.diskIds.add.includes(id)) {
    return true;
  }
  return false;
};

/** Function to check the delete button rendering */
const diskAction = (id, data) => {
  if (data.submitParams.diskIds.delete.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Delete'),
      text: __('Cancel Delete'),
      alt: __('Cancel Delete'),
      kind: 'ghost',
      callback: 'cancelDeleteDisk',
    };
  }
  return {
    is_button: true,
    title: __('Delete'),
    text: __('Delete'),
    alt: __('Delete'),
    kind: 'ghost',
    disabled: data.submitParams.diskIds.resize.includes(id),
    callback: 'deleteDisk',
  };
};

/** Function to check the resize button rendering */
const resizeAction = (id, data) => {
  if (data.submitParams.diskIds.resize.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Resize'),
      text: __('Cancel Resize'),
      alt: __('Cancel Resize'),
      kind: 'ghost',
      callback: 'cancelResizeDisk',
    };
  }
  return {
    is_button: true,
    title: __('Resize'),
    text: __('Resize'),
    alt: __('Resize'),
    kind: 'ghost',
    callback: 'resizeDisk',
    disabled: diskDisableButton(id, data),
  };
};

/** Function to validate the disk size
 * This function will validate the disk size in resize form
 */
export const validateDiskSize = (value, unit, minSize, minUnit) => {
  if (!Number.isNaN(value)) {
    const valueInMb = sizeInMB(value, unit);
    const minSizeInMb = sizeInMB(minSize, minUnit);
    if (valueInMb >= minSizeInMb) {
      return true;
    }
  }
  return false;
};

/** Function to set the Disk datatable row data
 * This formatting is needed for resize function validation
 */
export const formatDiskData = (disks) => {
  disks.forEach((disk) => {
    disk.orgHdSize = disk.hdSize;
    disk.orgHdUnit = disk.hdUnit;
  });

  return disks;
};

/** Function to set the Disk datatable row data  */
export const setDiskData = (disks, roles, data, setData) => {
  const diskData = [];
  disks.forEach((disk, i) => {
    const id = `disk${i}`;
    diskData[i] = {
      id,
      name: disk.hdFilename || ' ',
      type: disk.hdType,
      size: disk.hdSize,
      unit: disk.hdUnit,
      action: diskAction(id, data),
    };
    if (roles.isVmwareInfra) {
      diskData[i].dependent = disk.cb_dependent ? __('Yes') : __('No');
      diskData[i].mode = disk.hdMode || __('None');
      diskData[i].controller = disk.new_controller_type || __('None');
    }
    if (!roles.isVmwareCloud) {
      diskData[i].backing = {
        is_toggle: true,
        labelText: __('Delete Backing'),
        labelA: __('Yes'),
        labelB: __('No'),
        toggled: !!disk.delete_backing,
        ontoggle: () => backupDelete(id, data, setData),
        disabled: diskDisableButton(id, data),
      };
    }
    if (roles.isRedhat) {
      diskData[i].bootable = disk.cb_bootable === true ? __('Yes') : __('No');
    }
    if (roles.allowDiskSizeChange) {
      diskData[i].resize = resizeAction(id, data);
    }
  });
  return diskData;
};

/** Function to set the Disk datatable headers  */
export const getDiskTableHeaders = (roles) => {
  const headers = [
    { key: 'name', header: __('Name') },
    { key: 'type', header: __('Type') },
    { key: 'size', header: __('Size') },
    { key: 'unit', header: __('Unit') },
  ];
  if (roles.isVmwareInfra) {
    headers.push(
      { key: 'mode', header: __('Mode') },
      { key: 'controller', header: __('Controller Type') },
      { key: 'dependent', header: __('Dependent') }
    );
  }
  if (!roles.isVmwareCloud) {
    headers.push({ key: 'backing', header: __('Delete Backing') });
  }
  if (roles.isRedhat) {
    headers.push({ key: 'bootable', header: __('Bootable') });
  }
  if (roles.allowDiskSizeChange) {
    headers.push({ key: 'resize', header: __('Resize') });
  }
  headers.push({ key: 'action', header: __('Action'), contentButton: true });
  return headers;
};

/** function to execute disk sub forms submit gets called
 * Disk add, resize function
*/
export const setDiskFormSubmit = (data, setData, formData, roles, renderData) => {
  const dataTableData = data.dataTable.disks || [];
  if (data.form.action === 'add') {
    // code for add
    const id = `disk${dataTableData.length}`;
    const diskDatatableData = {
      id,
      hdFilename: '',
      hdType: formData.type,
      hdSize: formData.size,
      hdUnit: formData.unit,
      add_remove: data.form.action,
    };
    const submitDiskData = {
      id,
      disk_size_in_mb: sizeInMB(formData.size, formData.unit),
      thin_provisioned: formData.type === 'thin',
      type: formData.type,
    };
    if (roles.isVmwareInfra) {
      diskDatatableData.hdMode = formData.mode;
      diskDatatableData.new_controller_type = formData.controller;
      diskDatatableData.cb_dependent = formData.dependent || false;
      submitDiskData.persistent = formData.hdMode;
      submitDiskData.new_controller_type = formData.controller;
      submitDiskData.dependent = formData.dependent || false;
    }
    if (!roles.isVmwareCloud) {
      diskDatatableData.backing = false;
    }
    if (roles.isRedhat) {
      diskDatatableData.cb_bootable = formData.bootable || false;
      submitDiskData.bootable = formData.bootable || false;
    }
    data.submitParams.disks.add.push(submitDiskData);
    data.submitParams.diskIds.add.push(diskDatatableData.id);
    dataTableData.push(diskDatatableData);
  } else {
    // code for resize option
    const { id } = data.editingRow;
    const index = id.substring(4);
    dataTableData[index].old_size = dataTableData[index].hdSize;
    dataTableData[index].hdSize = formData.size;
    dataTableData[index].old_unit = dataTableData[index].hdUnit;
    dataTableData[index].hdUnit = formData.unit;
    const resizeData = {
      id,
      disk_name: dataTableData[index].hdFilename,
      disk_size_in_mb: sizeInMB(formData.size, formData.unit),
    };
    data.submitParams.disks.resize.push(resizeData);
    data.submitParams.diskIds.resize.push(id);
  }
  data.changeCounter += 1;
  setData({
    ...data,
    dataTable: {
      ...data.dataTable,
      disks: dataTableData,
    },
    form: {
      title: renderData.title,
      type: TYPES.RECONFIGURE,
      className: renderData.className,
    },
  });
};
