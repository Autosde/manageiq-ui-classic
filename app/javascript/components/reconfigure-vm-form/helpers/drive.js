import { TYPES } from './general';

/** Function to cancel the connection of the drive from device  */
export const cancelConnectDrives = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  datatableData[index].filename = datatableData[index].old_filename;

  const submitDrives = data.submitParams.drives.connect.filter((drive) => drive.id !== id);
  const connectedDriveIds = data.submitParams.driveIds.connect.filter((driveId) => driveId !== id);
  data.changeCounter -= 1;
  setData({
    ...data,
    isLoading: false,
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        connect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        connect: connectedDriveIds,
      },
    },
  });
};

/** Function to cancel the disconnection of drives from device  */
export const cancelDisconnectDrives = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  datatableData[index].filename = datatableData[index].disconnect_filename;

  const submitDrives = data.submitParams.drives.disconnect.filter((drive) => drive.id !== id);
  const disConnectedDriveIds = data.submitParams.driveIds.disconnect.filter((driveId) => driveId !== id);
  data.changeCounter -= 1;
  setData({
    ...data,
    isLoading: false,
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        disconnect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        disconnect: disConnectedDriveIds,
      },
    },
  });
};

/** Function to disconnect drives from device  */
export const disconnectDrives = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  const submitDrives = data.submitParams.drives.disconnect || [];
  const disConnectedDriveIds = data.submitParams.driveIds.disconnect || [];
  datatableData[index].disconnect_filename = datatableData[index].filename;
  datatableData[index].filename = '';
  const disconnectDriveData = {
    id,
    device_name: datatableData[index].device_name,
  };
  submitDrives.push(disconnectDriveData);
  disConnectedDriveIds.push(id);
  data.changeCounter += 1;
  setData({
    ...data,
    isLoading: false,
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        disconnect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        disconnect: disConnectedDriveIds,
      },
    },
  });
};

/** Function to check the connect button render options  */
const connectAction = (id, file, data) => {
  if (data.submitParams.driveIds.connect.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Connect'),
      text: __('Cancel Connect'),
      alt: __('Cancel Connect'),
      kind: 'ghost',
      callback: 'cancelConnectDrives',
    };
  }
  return {
    is_button: true,
    title: __('Connect'),
    text: __('Connect'),
    alt: __('Connect'),
    disabled: !!file,
    kind: 'ghost',
    callback: 'connectDrives',
  };
};

/** Function to check the disconnect button render options  */
const disConnectAction = (id, data) => {
  if (data.submitParams.driveIds.disconnect.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Disconnect'),
      text: __('Cancel Disconnect'),
      alt: __('Cancel Disconnect'),
      kind: 'ghost',
      callback: 'cancelDisconnectDrives',
      disabled: data.submitParams.driveIds.connect.includes(id),
    };
  }
  return {
    is_button: true,
    title: __('Disconnect'),
    text: __('Disconnect'),
    alt: __('Disconnect'),
    kind: 'ghost',
    callback: 'disconnectDrives',
    disabled: data.submitParams.driveIds.connect.includes(id),
  };
};

/** Function to set the CD/ROM Drive datatable row data  */
export const setDrivesData = (drives, data) => {
  const driveData = [];

  drives.forEach((drive, i) => {
    const id = `drive${i}`;
    driveData[i] = {
      id,
      name: drive.device_name,
      hostFile: drive.filename,
      action: connectAction(id, drive.filename, data),
    };
    if (drive.filename || drive.disconnect_filename) {
      driveData[i].disconnect = disConnectAction(id, data);
    } else {
      driveData[i].disconnect = '';
    }
  });

  return driveData;
};

/** Function to set the CD-ROM drive datatable headers  */
export const getDriveTableHeaders = () => [
  { key: 'name', header: __('Name') },
  { key: 'hostFile', header: __('Host File') },
  { key: 'disconnect', header: __('Disconnect') },
  { key: 'action', header: __('Actions') },
];

/** function to execute drive sub forms submit gets called
 * Drive connect form submission
*/
export const setDriveFormSubmit = (data, setData, formData, renderData) => {
  const { id } = data.editingRow;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  const submitDrives = data.submitParams.drives.connect || [];
  const connectedDriveIds = data.submitParams.driveIds.connect || [];
  const isoSelect = formData.host_file.split(',');

  datatableData[index].old_filename = datatableData[index].filename;
  // eslint-disable-next-line prefer-destructuring
  datatableData[index].filename = isoSelect[0];

  const connectDriveData = {
    id,
    device_name: datatableData[index].device_name,
    filename: datatableData[index].filename,
    storage_id: isoSelect[1],
  };
  submitDrives.push(connectDriveData);
  connectedDriveIds.push(id);
  data.changeCounter += 1;

  setData({
    ...data,
    isLoading: false,
    form: {
      title: renderData.title,
      type: TYPES.RECONFIGURE,
      className: renderData.className,
    },
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        connect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        connect: connectedDriveIds,
      },
    },
  });
};
