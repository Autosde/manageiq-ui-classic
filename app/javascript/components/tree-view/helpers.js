import { Tree } from 'react-wooden-tree';

/**
 * Helper
 * Changes class to className.
 * Sets undefined checkbox to null.
 * Moves the key to attr.key.
 */
const convert = (tree, fn) => tree.map((n) => {
  let node = { ...n };
  if (node.class) {
    node = { ...node, classes: node.class };
    delete node.class;
  }

  // The server cannot send null, just undefined, this is why it is tested as a string.
  if (node.state.checked === 'undefined') {
    node = { ...node, state: { ...node.state, checked: null } };
  } else if (node.state.checked === null) {
    node = { ...node, state: { ...node.state, checked: false } };
  }

  if (node.key) {
    node = { ...node, attr: { key: node.key } };
    delete node.key;
  }

  if (node.nodes) {
    node = { ...node, nodes: convert(node.nodes, fn) };
  }

  return fn ? fn(node) : node;
});

const flatten = data => Tree.convertHierarchicalTree(Tree.initHierarchicalTree(data));

/**
 * Helper
 * Activates the first active node in the tree, which is passed trought the props.
 *
 * FIXME Delete this function when bakcend is sending data where the initial node is activated,
 * or when all the trees calling a reducer if they want to have an activated node at the start.
 */

const activateNode = (tree, doActivate, key) => {
  if (!doActivate) {
    return tree;
  }

  let node = Tree.nodeSelector(tree, Tree.nodeSearch(tree, null, 'key', key)[0]);
  node = { ...node, state: { ...node.state, selected: true } };
  return { ...tree, [node.nodeId]: node };
};

const callBack = (nodeId, type, value, namespace) => ({ nodeId, type, value, namespace });

export { convert, flatten, activateNode, callBack };
