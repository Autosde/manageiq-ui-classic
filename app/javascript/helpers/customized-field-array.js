import React from 'react';
import { FieldArray } from '@@ddf';
import FormRenderer, { useFieldApi, useFormApi } from '@data-driven-forms/react-form-renderer';
import mapper from '../forms/mappers/componentMapper';

const ArrayItem = ({ fields, name }) => {
  const { renderForm } = useFormApi();
  const editedFields = fields.map((field) => ({ ...field, name: `${name}.${field.name}` }));

  return (
    <>
      {renderForm(editedFields)}
      <br />
    </>
  );
};

const FieldArrayCustom = (props) => {
  const {
    fieldKey, fields, itemDefault, ...rest
  } = useFieldApi(props);

  return (
    <FieldArray key={fieldKey} name={rest.input.name}>
      {(array) => (
        <>
          {array.fields.map((name, index) => (
            <ArrayItem
              key={`${name || fieldKey}`}
              fields={fields}
              name={name}
              fieldKey={fieldKey}
              fieldIndex={index}
            />
          ))}
          {(array.fields.length === 0) && array.fields.push(itemDefault)}
        </>
      )}
    </FieldArray>
  );
};

const componentMapper = {
  ...mapper,
  'custom-field-array': FieldArrayCustom,
};

const CustomizedFieldArray = async(arrayName, arrayFields) => {
  const getFields = (arrayFields) => arrayFields.map((field) => ({
    component: field.component,
    id: field.id,
    name: field.name,
    label: field.label,
    options: field.options,
  }));

  const schema = {
    fields: [
      {
        component: 'custom-field-array',
        name: arrayName,
        fieldKey: 'field_array',
        title: arrayName,
        fields: getFields(arrayFields),
      },
    ],
  };

  return (
    <>
      <FormRenderer
        componentMapper={componentMapper}
        schema={schema}
      />
    </>
  );
};

export { CustomizedFieldArray };
