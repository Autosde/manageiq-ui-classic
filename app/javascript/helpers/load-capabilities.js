import Api from '../http_api/api';

// in providers capabilities jsonb field, capabilities are saved like {'abstract_capability': 'name', 'uuid': '999..', 'value': 'true/false'},
// but in the storage_manager models the capability name is saved under different keys, instead of 'abstract_capability'.
// the parameter fieldName enables to dynamically access the correct key.

const loadCapabilities = (sourceType, sourceId, fieldName) => Api.get(`/api/${sourceType}/${sourceId}?attributes=capabilities`)
  .then(({ capabilities }) => capabilities.map((capability) => (
    { value: capability.uuid, label: `${capability[`${fieldName}`]}: ${capability.value}` })));

export default loadCapabilities;
