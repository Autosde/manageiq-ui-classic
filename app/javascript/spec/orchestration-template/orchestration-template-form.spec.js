import React from 'react';
import { act } from 'react-dom/test-utils';
import fetchMock from 'fetch-mock';

import '../helpers/addFlash';
import '../helpers/miqFlashLater';
import '../helpers/miqSparkle';
import '../helpers/miqAjaxButton';
import '../helpers/sprintf';
import { mount } from '../helpers/mountForm';

import OrcherstrationTemplateForm from '../../components/orchestration-template/orcherstration-template-form';
import miqRedirectBack from '../../helpers/miq-redirect-back';

const { default: CodeEditor } = jest.requireActual('../../components/code-editor');

describe('OrcherstrationTemplate form', () => {
  let initialProps;
  const sparkleOnSpy = jest.spyOn(window, 'miqSparkleOn');
  const sparkleOffSpy = jest.spyOn(window, 'miqSparkleOff');
  const addFlashSpy = jest.spyOn(window, 'add_flash');

  beforeEach(() => {
    initialProps = {
      managers: [['foo', 'bar']],
    };
  });

  afterEach(() => {
    fetchMock.reset();
    sparkleOnSpy.mockRestore();
    sparkleOffSpy.mockRestore();
    addFlashSpy.mockRestore();
  });

  it('should call submit function', async(done) => {
    fetchMock.postOnce('/api/orchestration_templates', {});
    let wrapper;
    await act(async() => {
      wrapper = mount(<OrcherstrationTemplateForm {...initialProps} />);
    });
    wrapper.update();

    await act(async() => {
      wrapper.find('input[name="name"]').simulate('change', { target: { value: 'foo' } });
      /**
         * manually change content value
         * Code component is not standard input element
         * Two first parameters are codemirror element and data
         */
      wrapper.find(CodeEditor).find('Controlled').props().onChange(null, null, 'Some random content');
      wrapper.find('form').simulate('submit');
    });

    expect(fetchMock.lastCall()).toBeTruthy();
    expect(JSON.parse(fetchMock.lastCall()[1].body)).toEqual(expect.objectContaining({
      name: 'foo',
      content: 'Some random content',
    }));
    expect(sparkleOnSpy).toHaveBeenCalledTimes(1);
    done();
  });

  it('should call miqFlashLater on cancel action', async(done) => {
    let wrapper;
    await act(async() => {
      wrapper = mount(<OrcherstrationTemplateForm {...initialProps} />);
    });
    wrapper.update();

    await act(async() => {
      wrapper.find('button.bx--btn--secondary').first().simulate('click');
    });

    expect(miqRedirectBack).toHaveBeenCalledWith(
      'Creation of a new Orchestration Template was cancelled by the user',
      'success',
      '/catalog/explorer',
    );
    done();
  });

  it('should render edit variant', async(done) => {
    const sparkleOnSpy = jest.spyOn(window, 'miqSparkleOn');
    fetchMock.patchOnce('/api/orchestration_templates/123', {});
    fetchMock.getOnce('/api/orchestration_templates/123?attributes=name,description,type,ems_id,draft,content', {
      name: 'foo',
      content: 'content',
    });
    let wrapper;
    await act(async() => {
      wrapper = mount(<OrcherstrationTemplateForm {...initialProps} otId={123} />);
    });
    wrapper.update();
    wrapper.find('input[name="name"]').simulate('change', { target: { value: 'bar' } });
    wrapper.update();
    await act(async() => {
      wrapper.find('form').simulate('submit');
    });

    expect(fetchMock.lastCall()).toBeTruthy();
    expect(JSON.parse(fetchMock.lastCall()[1].body)).toEqual(expect.objectContaining({
      name: 'bar',
      content: 'content',
    }));
    expect(sparkleOnSpy).toHaveBeenCalledTimes(1);
    done();
  });

  it('should render copy variant', async(done) => {
    const sparkleOnSpy = jest.spyOn(window, 'miqSparkleOn');
    fetchMock.postOnce('/api/orchestration_templates/123', {});
    fetchMock.getOnce('/api/orchestration_templates/123?attributes=name,description,type,ems_id,draft,content', {
      name: 'foo',
      content: 'content',
    });

    let wrapper;
    await act(async() => {
      wrapper = mount(<OrcherstrationTemplateForm {...initialProps} otId={123} copy />);
    });
    wrapper.update();

    await act(async() => {
      wrapper.find('input[name="name"]').simulate('change', { target: { value: 'bar' } });
      /**
           * manually change content value
           * Code component is not standard input element
           * Two first parameters are codemirror element and data
           */
      wrapper.find(CodeEditor).find('Controlled').props().onChange(null, null, 'updated content');
      wrapper.find('form').simulate('submit');
    });

    expect(fetchMock.lastCall()).toBeTruthy();
    expect(JSON.parse(fetchMock.lastCall()[1].body)).toEqual(expect.objectContaining({
      action: 'copy',
      resource: {
        name: 'bar',
        content: 'updated content',
      },
    }));
    expect(sparkleOnSpy).toHaveBeenCalledTimes(1);
    done();
  });
});

describe('Orcherstration Stack form', () => {
  let initialProps;
  let submitSpyMiqSparkleOn;
  let submitSpyMiqSparkleOff;
  let spyMiqAjaxButton;

  beforeEach(() => {
    initialProps = {
      managers: [['foo', 'bar']],
    };
    submitSpyMiqSparkleOn = jest.spyOn(window, 'miqSparkleOn');
    submitSpyMiqSparkleOff = jest.spyOn(window, 'miqSparkleOff');
    spyMiqAjaxButton = jest.spyOn(window, 'miqAjaxButton');
  });

  afterEach(() => {
    fetchMock.reset();
    submitSpyMiqSparkleOn.mockRestore();
    submitSpyMiqSparkleOff.mockRestore();
    spyMiqAjaxButton.mockRestore();
  });

  it('should render copy variant', async(done) => {
    const sparkleOnSpy = jest.spyOn(window, 'miqSparkleOn');
    fetchMock.postOnce('/api/orchestration_templates/123', {});
    fetchMock.getOnce('/api/orchestration_templates/123?attributes=name,description,type,ems_id,draft,content', {
      name: 'foo',
      content: 'content',
    });

    let wrapper;
    await act(async() => {
      wrapper = mount(<OrcherstrationTemplateForm {...initialProps} isStack otId={123} copy />);
    });
    wrapper.update();

    await act(async() => {
      wrapper.find('input[name="name"]').simulate('change', { target: { value: 'bar' } });
      /**
           * manually change content value
           * Code component is not standard input element
           * Two first parameters are codemirror element and data
           */
      wrapper.find(CodeEditor).find('Controlled').props().onChange(null, null, 'updated content');
      wrapper.find('form').simulate('submit');
    });

    expect(fetchMock.lastCall()).toBeTruthy();
    expect(sparkleOnSpy).toHaveBeenCalledTimes(1);
    expect(wrapper).toMatchSnapshot();
    done();
  });

  it('should call submit function', async(done) => {
    const addContent = {
      templateId: 123,
      templateName: 'template_name',
      templateDescription: 'template_description',
      templateDraft: 'true',
      templateContent: 'template_content',
    };

    miqAjaxButton(`/orchestration_stack/stacks_ot_copy?button=add`, addContent);
    expect(spyMiqAjaxButton).toHaveBeenCalledWith('/orchestration_stack/stacks_ot_copy?button=add', addContent);
    done();
  });

  it('should call cancel function', async(done) => {
    const cancelMessage = __('Copy of Orchestration Template was cancelled by the user');
    miqRedirectBack(cancelMessage, 'success', `/orchestration_stack/show/${123}?display=stack_orchestration_template#/`);

    expect(miqRedirectBack).toHaveBeenCalledWith(
      cancelMessage, 'success', `/orchestration_stack/show/${123}?display=stack_orchestration_template#/`
    );
    done();
  });
});
