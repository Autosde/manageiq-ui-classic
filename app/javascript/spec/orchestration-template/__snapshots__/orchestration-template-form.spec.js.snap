// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orcherstration Stack form should render copy variant 1`] = `ReactWrapper {}`;
